### ISSUE TITLE
Coercion accepted but not used

### ISSUE THREAD

**Issue by dselsam** [leanprover/lean3#644]
Hi,

I am having trouble coercing bool-valued predicates to Prop-valued predicates. I am able to coerce bool to Prop:

```
definition bool_to_prop [coercion] (b : bool) : Prop := b = tt
definition decidable_bool_to_prop [instance] : decidable_pred bool_to_prop :=
  bool.rec (decidable.inr bool.no_confusion) (decidable.inl rfl)

definition bpred [reducible] (T : Type) := T → bool
definition ppred [reducible] (T : Type) := T → Prop

definition negb : bpred bool := bool.rec tt ff

eval if negb tt then 1 else 222
```

I can register a coercion from `bpred` to `ppred` in two different ways:

```
definition comp_to_pred1 [coercion] {T : Type} (comp : bpred T) : ppred T := 
  λ x, bool_to_prop (comp x)
definition comp_to_pred2 [coercion] {T : Type} (comp : bpred T) : T → Prop:= 
  λ x, bool_to_prop (comp x)
```

But it doesn't seem to work when trying to use a `bpred` to filter a list:

```
eval @list.filter bool negb _ list.nil
```

gives the error: "term `negb` has type `bpred bool` but is expected to have type `bool -> Prop`"

Thanks in advance for any help.

**Comment by leodemoura** [comment]
There is a missing feature that is the source of the problem. Lean is not "lifting" coercions automatically for us yet. That is, if we have a coercion from `A` to `B`, Lean should automatically infer one from `C -> A` to `C -> B`, and from `B -> C` to `A -> C`. After we implement this feature, your example will work smoothly. You will only need to define the coercion from `bool` to `Prop`, 

In the current version, I think the simplest workaround is to define `to_pred` and `to_pred_dec` once:

``` lean
definition to_pred {A : Type} (p : A → bool) : A → Prop :=
λ a, p a = tt

definition to_pred_dec [instance] {A : Type} (p : A → bool) : decidable_pred (to_pred p) :=
begin
  intro a, unfold to_pred,
  eapply (bool.rec_on (p a)),
    right, contradiction,
    left, reflexivity
end
```

and then explicitly use `to_pred` to convert `A -> bool` into `A -> pred`. Example:

``` lean
definition negb : bool → bool := bool.rec tt ff
eval filter (to_pred negb) [tt, ff, tt, ff]
```

BTW, we also recently changed how coercions are used (see #547).
This thread may be useful. The main issue is how Lean detects whether a coercion is needed or not. Lean allow us to define coercions from types of the form `(C ...)`, where `C` is a constant.
In the new approach, if we have a type `t`, Lean will normalize it, but it will assume that any constant `C`, which is a source/target of a coercion, is opaque (i.e., it is not unfolded).
So, if we have a coercion from `bpred T` to `ppred T`, it will only be applied if our type `t` is a `bpred T` or it normalizes to `bpred T`. So, the coercion will not be used if `t` is `T -> bool`. 
Note that, the following works:

``` lean
definition bpred (T : Type) := T → bool
definition ppred (T : Type) := T → Prop

definition my_negb : bpred bool := negb

definition bool_to_prop  [coercion] (b : bool) : Prop := b = tt
definition comp_to_pred1 [coercion] {T : Type} (comp : bpred T) : ppred T :=
λ x, bool_to_prop (comp x)

variable f : ppred bool → Prop
check f my_negb
```

However, here `my_negb` is a `bpred bool` and `f` is expecting a `ppred bool`.

**Comment by leodemoura** [comment]
Remark: I did not mark this issue as a bug, because a coercion from `bpred T` is not supposed to be considered when the given type is `T -> bool` (see discussion at #540). I marked it as an enhancement because it will a nonissue after we implement "coercion lifting"

**Comment by leodemoura** [comment]
I implemented "coercion lifting". Now we can write:

``` lean
import data.list
open bool list

definition bool_to_Prop [coercion] [reducible] (b : bool) : Prop := b = tt

definition bpred_dec [instance] {A : Type} (p : A → bool) : ∀ a, decidable (p a = tt) :=
begin
  intro a,
  eapply (bool.rec_on (p a)),
    right, contradiction,
    left, reflexivity
end

definition negb : bool → bool := bool.rec tt ff

example : filter negb [tt, ff, tt, ff] = [ff, ff] :=
rfl
```

**Comment by dselsam** [comment]
Thank you, this is fantastic.

**Comment by htzh** [comment]
Sometimes the coercion happens a bit too early which makes it problematic for overloading resolution. For example see:
https://github.com/htzh/leanproved/blob/a7d4337f579dc17a89d242fc56a797ad2e30385c/group_theory/action.lean#L67
If I remover perm.f it complains about '*'.

**Comment by leodemoura** [comment]
@htzh A coercion is not considered in this case. 
Here is the code that decides whether a coercion to function-class is considered or not.
https://github.com/leanprover/lean/blob/master/src/frontends/lean/elaborator.cpp#L422
Note the following line:
https://github.com/leanprover/lean/blob/master/src/frontends/lean/elaborator.cpp#L431
If the type of the term is "stuck", then no coercion to the function class is considered.
We say a term is stuck if it is of the form
- `?M ...`  where ?M is a metavariable (i.e., a term that must be synthesized); OR
- `C.rec ... t ...` where `C.rec` is an eliminator and `t` is a stuck term

In principle, we could add a case-split, but it would increase the state space considerably since whenever we have a stuck term where a function is expected, we would have to inject a case-split. 

In your definition, the type of `(hom f) * (hom g)` is stuck because of the overloading.

Another alternative we considered was to propagate constraints (without performing any case-split), and then decide whether coercions should be considered or not. I believe this would solved this instance. However, I'm concerned it would be trickier to understand what is going on when the elaborator fails.

Possible workarounds:
- provide the coercion explicitly (like you did)
- provide explicit type information: 

``` lean
          hom (f*g) a = ((hom f) * (hom g) : perm S) a : is_hom hom
```
- define local notation with type "hints"

``` lean
local infix `×` := @has_mul.mul (perm S) _
...
        hom (f*g) a = ((hom f) × (hom g)) a : is_hom hom
```

**Comment by htzh** [comment]
I will try to understand the code sections you linked to. I came to the understanding that there was coercion due to the following error message if I remove perm.f projection:

```
failed to synthesize placeholder
G : Type,
S : Type,
ambientG : group G,
finS : fintype S,
deceqS : decidable_eq S,
deceqG : decidable_eq G,
hom : G → perm S,
H : finset G,
a : S,
Hom : is_hom_class hom,
f g : G,
Pgstab : g ∈ stab hom H a,
Pg : perm.f (hom g) a = a
⊢ has_mul (S → S)
\Users\haitao\proj\lean\proved\group_theory\flycheck_action.lean:67:31: previous error additional information
none of the overloads is applicable: algebra.has_mul.mul finset.product
```

The message is saying that Lean could not infer has_mul for (S → S), which is the type of the coerced function. So it appears to me that Lean has coerced (hom f) and (hom g) to (S → S) and then tried to look for an instance of has_mul for the function type and failing that reported that overload resolution has failed. That is what I meant when I said "sometimes the coercion happens a bit too early ... "

If you are saying that after it is stuck no coercion is considered I understand and agree. But I don't understand if you meant no coercion is applied here because without the coercion the overload resolution should succeed.

**Comment by leodemoura** [comment]
> If you are saying that after it is stuck no coercion is considered I understand and agree. But I don't understand if you meant no coercion is applied here because without the coercion the overload resolution should succeed.

I'm talking about the coercion `perm.f` from `perm S` to the "function class".
This is the coercion that is ignored.

**Comment by leodemoura** [comment]
To complement: it is ignored because the type of `(hom f) * (hom g)` is stuck at preprocessing time.

**Comment by htzh** [comment]
So it sounds like that Lean: 1) determines type of `a` to be S; 2) infers `(hom f) * (hom g)` should have type S -> S; 3) gets stuck because no \* exists for S -> S.

Alternatively one could: 1) determine `hom f` to be of type perm S; 2) infer \* from the group instance of perm S; 3)  infer `(hom f) * (hom g)` to be perm S; 4) infer that coercion is needed from perm S to S -> S

So is this the alternative you were referring to? (it is what I had in mind):

> Another alternative we considered was to propagate constraints (without performing any case-split), and then decide whether coercions should be considered or not. I believe this would solved this instance. However, I'm concerned it would be trickier to understand what is going on when the elaborator fails.

I would agree that the situation is complicated here with three features interacting: (perm S) group class instance inference, \* overloading resolution, and perm.f coercion to function. It can only succeed if the above are applied in that order.

**Comment by leodemoura** [comment]
> So is this the alternative you were referring to?

None of them. It is sad, but the error message content cannot be used to understand what Lean did in this case. The decision whether a coercion should be used or not was made before it obtained all the information that is reflected in the error message.

The Lean elaboration procedure has two main steps: preprocessing and constraint solving.
I'm ignoring the interaction with tactics here.
In the preprocessing step, we convert a "preterm" (what was typed by the user) into a term containing metavariables and a set of constraints. The metavariables correspond to what must be synthesized.
The preprocessor decides whether a coercion must be considered in a particular place or not.
For example, suppose we have a term `t`. Here are possible scenarios:

1- "easy case" suppose `t` has type `C ...` but it is being used in a context where a `D ...` is expected. Moreover, we have a single coercion `f` from `C ...` to `D ...`. Then, the preprocessor replaces `t` with `f t`. 

2- Suppose there are multiple coercions `f_1`, ..., `f_n`, from `C ...` to `D ...`. Then the preprocessor creates a fresh metavariable ?M, adds the constraint `?M in {f_1 t, ..., f_n t}`, and replaces `t` with `?M`. We are using the constraint to "postpone" the choice to the constraint solver.

3- Suppose `t`'s type is "stuck" term (e.g., `?Mt ...`). That is, we don't really know what the type of `t` is before we start solving other constraints that have been generated. If `D ...` there are coercions to `D ...`, we create a fresh metavariable `?M`, a constraint `?M in {t, f_1 t, f_2 t, ...}`. The first one corresponds to the case where no coercion is needed. Actually, the constraint is of the form `?M in fn` where `fn` is a function that enumerates all possible alternatives. The constraint solver also provides to `fn` the type of `t` at constraint solving time. 

As I said in the previous comments, case 3 is only considered when `D` is a constant.  Lean ignores  case 3 when `D ...` is a Pi. This is why your example fails. We made this decision in an attempt to reduce the search space.

BTW, there are more cases, I tried to provide a sample of them in the list above.
I'm also ignoring the "coercion lifting" that is a new feature added for solving issue #644.

Regarding the error message, it is produced by the constraint solver when it fails to solve the constraints generated by the preprocessing step. Note that the terms produced by the preprocessor are full of metavariables and very hard to understand. So, we apply the partial solution (an assignment from metavariables to terms) produced by the constraint solver before displaying the error. In general, it helps, in this case, it generated more confusion :-(

To address the issue in your file, the natural solution is to consider case 3 even when `D ...` is a Pi.

BTW, there are other cases we do not consider. 
Example: if `t`'s type and the expected type are "stuck" terms.
